### PR TITLE
clippy

### PR DESCRIPTION
**Pull request by mightyiam**
Co-authored-by: warren2k <846021+warren2k@users.noreply.github.com>

### CODE REVIEW

**Comment by mightyiam** [comment]
This is based on top of #720 . I suggest reviewing that one, first.

**Comment by mightyiam** [comment]
Rebased. Please approve workflow and review. Reminder: this is based on top of #720.

**Comment by mightyiam** [comment]
Our last workflow run failed. Force-pushed an attempt to resolve the issue. Please approve another workflow run.

File: .github/workflows/ci.yml
Comment (jswrenn) [inline .github/workflows/ci.yml:38]: Let's only run clippy on the msrv toolchain. Clippy recommendations that violate msrv are not actionable.

File: wrappers.rs
Comment (jswrenn) [inline wrappers.rs:38]: As I wrote on #720, I'm not keen on the wrapper pattern to suppress warnings.

**Review by jswrenn — CHANGES_REQUESTED** [review]

**Comment by mightyiam** [comment]
Could you please approve workflow? This isn't ready, though.

File: .github/workflows/ci.yml
Comment (mightyiam) [reply .github/workflows/ci.yml]: @jswrenn again, we recommend option B. Call it and we'll do it.

**Comment by jswrenn** [comment]
Alright, let's do clippy on stable. I don't like the idea that our CI might spontaneously break upon new Rust releases, but hopefully that's rare.

**Review by Philippe-Cholet — APPROVED** [review]

**Comment by mightyiam** [comment]
Looks like conflicts. I'll see what I can do.

**Comment by Philippe-Cholet** [comment]
I don't know if we will apply CI changes, I leave that decision to jswrenn but I sure would like to fix lints.

**Comment by jswrenn** [comment]
Whoops, I must have missed when this PR was marked ready for review. I'm still fine running clippy with the stable toolchain.

File: benches/bench1.rs
Comment (mightyiam) [inline benches/bench1.rs:479]: I'm not sure what this benchmark is about. I'm only guessing that calling enumerate and then ignoring the index is intentional. Same goes for the similar ones below.

File: src/either_or_both.rs
Comment (mightyiam) [inline src/either_or_both.rs:497]: Does anyone know why we're implementing Into instead of From?

File: src/multipeek_impl.rs
Comment (mightyiam) [inline src/multipeek_impl.rs:82]: https://rust-lang.github.io/rust-clippy/master/index.html#/get_first

https://doc.rust-lang.org/std/collections/struct.VecDeque.html#method.front

File: src/tuple_impl.rs
Comment (mightyiam) [inline src/tuple_impl.rs:65]: https://rust-lang.github.io/rust-clippy/master/index.html#/unnecessary_lazy_evaluations

File: src/either_or_both.rs
Comment (mightyiam) [inline src/either_or_both.rs:32]: https://rust-lang.github.io/rust-clippy/master/index.html#/match_like_matches_macro

File: tests/quick.rs
Comment (mightyiam) [inline tests/quick.rs:783]: https://rust-lang.github.io/rust-clippy/master/index.html#/explicit_counter_loop

File: tests/quick.rs
Comment (mightyiam) [inline tests/quick.rs:1001]: https://rust-lang.github.io/rust-clippy/master/index.html#/redundant_pattern_matching

File: tests/test_std.rs
Comment (mightyiam) [inline tests/test_std.rs:43]: https://rust-lang.github.io/rust-clippy/master/index.html#/useless_conversion

File: tests/test_std.rs
Comment (mightyiam) [inline tests/test_std.rs:1268]: https://rust-lang.github.io/rust-clippy/master/index.html#/non_canonical_partial_ord_impl

**Review by mightyiam — COMMENTED** [review]
Rebased and resolved 1.75.0 lints.

File: src/either_or_both.rs
Comment (Philippe-Cholet) [reply src/either_or_both.rs:497]: I don't see a point of not doing it so I suggest you convert it to a `From` implementation in a separate commit.

**Review by Philippe-Cholet — APPROVED** [review]
Thanks for updating your work.

File: src/either_or_both.rs
Comment (mightyiam) [reply src/either_or_both.rs:497]: Here's [an issue](https://github.com/rust-itertools/itertools/issues/836) for that.

Now that shouldn't block this PR?

Comment (Philippe-Cholet) [reply src/either_or_both.rs:497]: I sure want this PR to be merged as soon as possible as you made quite a work here!

**Comment by Philippe-Cholet** [comment]
So we chose "Option B" from your https://github.com/rust-itertools/itertools/pull/740#discussion_r1335205248 in which you mention something interesting.

> When bumping MSRV, go through all the allowed clippy lints in the codebase and examine which ones can now be acted upon.

I think we might easily forgot that so I think that we should have a comment on the "rust-version" line in "Cargo.toml".
Once done, I'll merge this PR as jswrenn's last comment is enough confirmation to me.

File: src/either_or_both.rs
Comment (bishopcheckmate) [inline src/either_or_both.rs:32]:
```suggestion
        matches!(self, Left(_))
```

File: src/either_or_both.rs
Comment (bishopcheckmate) [inline src/either_or_both.rs:38]:
```suggestion
        matches!(self, Right(_))
```

**Review by bishopcheckmate — COMMENTED** [review]
nit, actually dereferencing is not needed

File: src/either_or_both.rs
Comment (Philippe-Cholet) [reply src/either_or_both.rs:32]: Indeed, it compiles to the exact same thing (I just checked a simple example on godbolt).
I'm not sure we should care when I did not find any clippy lint for it.
I'll let the author decide here.

Comment (bishopcheckmate) [reply src/either_or_both.rs:32]: yeah, that's definitely a nitpick. on the other hand in a pr which satisfies clippy, it might be kind of future proofing :smile:

Comment (mightyiam) [reply src/either_or_both.rs]: Sorry thank you done.

File: Cargo.toml
Comment (mightyiam) [inline Cargo.toml:19]: Added this here. Good enough?

**Review by Philippe-Cholet — APPROVED** [review]
This is it, thanks for your dedicated work. 🎉

File: benches/bench1.rs
Comment (Philippe-Cholet) [reply benches/bench1.rs:479]: @mightyiam After merging this PR, I search for PRs/issues about clippy (and closed some), saw #618 and eventually ran `cargo clippy --all-targets` and got `clippy::unused_enumerate_index` as "unknown lint" (yet I found it online). _What am I missing here?_
And second thing, we don't run clippy with `--all-target` and I'm wondering why, I don't see any discussion about it.

Comment (mightyiam) [reply benches/bench1.rs:479]: I think we run clippy four times because of a matrix value `features`. One of the is `--all-targets --all-features`.

What version toolchain did you get an error with and what is the error exactly, please?

Comment (Philippe-Cholet) [reply benches/bench1.rs:479]: Ahh! [unused_enumerate_index](https://rust-lang.github.io/rust-clippy/master/index.html#/unused_enumerate_index) is added in 1.75 and I'm still on 1.74, my bad sorry! I'm so used on thinking I can't use recent features of Rust I did not thought it could be such a recent lint.
And another mistake of mine on matrix features. Ok, thanks!

**Pull request merged**
